src/main/java/org/jboss/modules/PathUtils.java

```diff
@@ -268,21 +268,41 @@ static String directCanonicalize(String path) {
     }
 
 
+    private static boolean isSepChar(char ch) {
+        // on non-Windows, the JIT *should* eliminate the second check as redundant
+        return ch == '/' || ch == File.separatorChar;
+    }
+
     /**
      * Canonicalize the given path.  Removes all {@code .} and {@code ..} segments from the path.
      *
      * @param path the relative or absolute possibly non-canonical path
      * @return the canonical path
      */
     public static String canonicalize(String path) {
-        if (path.trim().isEmpty()) {
+        // is null check here even needed?
+        if (path == null || path.isEmpty()) {
             return path;
         }
 
-        // represents a state transitions that will be found via the state machine
-        // but if these patterns aren't contained in the string they can't be found
-        // via the state machine. (we know it already contains at least "." at this point.)
-        if (!path.contains("/.") && !path.contains("./")) {
+        boolean found = false;
+        int idx = path.indexOf('.'); // maybe reuse the result of the above comparison for efficiency
+        // shortcut if string is "." or ".."
+        if (idx == 0 && path.length() == 1 || path.length() == 2 && path.charAt(idx+1) == '.') {
+            return "";
+        }
+        if (idx == -1) {
+            return path;
+        }
+        do {
+            // represents a state transitions that will be found via the state machine
+            // but if these patterns aren't contained in the string they can't be found
+            // via the state machine. (we know it already contains at least "." at this point.)
+            if (path.length() > idx + 1 && isSepChar(path.charAt(idx + 1)) || idx > 0 && isSepChar(path.charAt(idx - 1))) {
+                found = true;
+            }
+        } while (idx + 1 < path.length() && (idx = path.indexOf('.', idx + 1)) != -1);
+        if (!found) {
```

Comment:
I like this less and less the more complex it gets. I think it misses strings like `/` and `//` and `foo//bar` as well, while duplicating a lot of the work of the state machine.

The right fix is probably to drop all of this stuff and modify the state machine to only call `getBuffer` and initially populate the buffer once it knows that the string will need modifications (e.g. when encountering a `/` in states 1, 2, or 3), and then having the emit routine at the bottom skip emitting unless the buffer is not null. Maybe also replacing the `switch` statements with more optimal `if` trees, which the JIT can profile much better.